### DISCUSSION TITLE
How do I add 2 auxiliary chains to the pool?

### DISCUSSION THREAD

**Discussion by pepecoinppc** [dreams-money/merged-mining-pool#6] [Q&A]
Hi! This project is awesome!

It seems like it's a little bit lacking in documentation though. I was wondering how you add a 2nd AuxPow coin. I'm able to mine Litecoin and Dogecoin on the testnets, but not a 3rd coin. I'm trying to add Pepecoin (a fork of Dogecoin). I added a `pepecoin.go` class, added it to the `chain.go` switch statement, but it's still not submitting blocks to Aux2.

I have a hunch that's it's not quite complete yet. Has support for multiple aux chains not been implemented yet?

Thanks!

**Comment by dreams-money** [comment]
Hello, pepecoinppc!

I appreciate your interest in the project.   You're correct in explaining that multiple aux chains are not supported, and I've been unable to find the appropriate documentation to achieve multiple aux chains.

In the docs/ folder, you'll find the best notes I've seen on the topic.  I've noted fields that require further research. Block assembly and serialization knowledge is helpful to further the research.

If you're familiar with Dogecoin's source code, another breadcrumb would be their "createauxblock" RPC method - Dogecoin creates the auxiliary data for you.

I'm happy to provide you with what I have so far.  Adding multiple aux coins would be extremely helpful.

**Reply by pepecoinppc** [reply]
That would be great! Just from a developer's perspective, there are no easy to set up pools that support merge mining afaik. I would like to actually see Pepecoin being merge mined with Litecoin and Dogecoin before I do a release you know? 

I stumbled across this on github the other day, it may be of some interest to you https://github.com/TheSuperiorCoin/pool-merge-coin-mining/blob/master/pool-merge-coin-mining.md

Perhaps we can ask Patrick Lodder for some insight?

**Reply by pepecoinppc** [reply]
Something else I would recommend is tagging the project with tags like these `merge-mining`, `dogecoin`, `litecoin`, `mining`, `scrypt`, `blockchain`, `pool`, `mining-pool`, `mining-pools`, etc. This will make it easier for people to find it. This is probably one of the best pools on github tbh.

**Comment by dreams-money** [comment]
Thank you!  I agree that there are no other merged mining pools at the moment.  It took me 2 years to fully grasp merged mining and develop this pool.  I'd love to support Pepecoin if time allows.

Also, thank you for that resource.  I'll digest it when the time comes.

Patrick may be the best resource to understand the block serialization of a 1 > aux coin mined block.   I think we'd need to identify a block on the chain that had been mined as such.  Bitcoin has plenty of examples.

**Reply by pepecoinppc** [reply]
I can help with identifying merge mined blocks. Mining Dutch is currently merge mining lots of scrypt coins. They also label themselves in the coinbase so you know for sure they mined the block. Here's a list of coins they're merge mining right now.

This is on mining-dutch's website

![image](https://github.com/dreams-money/merged-mining-pool/assets/143240103/b61a923c-f851-42ad-bc4e-3edd652c9bfe)

Here on MiningPoolStats, blocks are labeled Mining Dutch if they were mined by that pool. Here you can see stats for Dingocoin. The list of blocks at the bottom of the page tell you who mined the block.

https://miningpoolstats.stream/dingocoin

It looks like block 1551261 was mined by Mining Dutch

![image](https://github.com/dreams-money/merged-mining-pool/assets/143240103/9ae77698-870b-4dbb-a6a8-2fdf84ba2dec)

Here's the block when grabbed from the dingocoin-cli "getblock" rpc function.

```
{
  "hash": "0963a27286c8b43faed5754847395fd4514d18e8a7d5c5a8f2a441489b668f33",
  "confirmations": 7,
  "strippedsize": 1094,
  "size": 1094,
  "weight": 4376,
  "height": 1551261,
  "version": 3277060,
  "versionHex": "00320104",
  "merkleroot": "a75cc673c2f6638b17f430c3a04bded2882616133c648f1868c887b48fad6c11",
  "tx": [
    "a75cc673c2f6638b17f430c3a04bded2882616133c648f1868c887b48fad6c11"
  ],
  "time": 1714164618,
  "mediantime": 1714164379,
  "nonce": 0,
  "bits": "1a301aee",
  "difficulty": 348755.6867141065,
  "chainwork": "00000000000000000000000000000000000000000000001ccb3e3dd3ba8c6777",
  "auxpow": {
    "tx": {
      "hex": "01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff5203aad028048e132c6608fabe6d6da4251cead87fcae154b8a36c2dc68859ae91fdd69cee4a21e9f573f9ec2727c10001000000000000024e952c320600004300112f4d696e696e672d44757463682f2d3337000000000213836925000000001976a91466b4153493a3034dd256ef416ac7ed2c1b1b310d88ac0000000000000000266a24aa21a9ed3205f889713b13abee37da0e42655405a1c4e727949074a1a5a06bdc4f4fec9300000000",
      "txid": "c9127f238697a6e466556c65364114904d019f140d382ccde14b1b12840659a7",
      "hash": "c9127f238697a6e466556c65364114904d019f140d382ccde14b1b12840659a7",
      "size": 214,
      "vsize": 214,
      "version": 1,
      "locktime": 0,
      "vin": [
        {
          "coinbase": "03aad028048e132c6608fabe6d6da4251cead87fcae154b8a36c2dc68859ae91fdd69cee4a21e9f573f9ec2727c10001000000000000024e952c320600004300112f4d696e696e672d44757463682f2d3337",
          "sequence": 0
        }
      ],
      "vout": [
        {
          "value": 6.27671827,
          "n": 0,
          "scriptPubKey": {
            "asm": "OP_DUP OP_HASH160 66b4153493a3034dd256ef416ac7ed2c1b1b310d OP_EQUALVERIFY OP_CHECKSIG",
            "hex": "76a91466b4153493a3034dd256ef416ac7ed2c1b1b310d88ac",
            "reqSigs": 1,
            "type": "pubkeyhash",
            "addresses": [
              "DEW9CCYHNpWzuyFWcjkFj2CsbLwiWifw1c"
            ]
          }
        },
        {
          "value": 0.00000000,
          "n": 1,
          "scriptPubKey": {
            "asm": "OP_RETURN aa21a9ed3205f889713b13abee37da0e42655405a1c4e727949074a1a5a06bdc4f4fec93",
            "hex": "6a24aa21a9ed3205f889713b13abee37da0e42655405a1c4e727949074a1a5a06bdc4f4fec93",
            "type": "nulldata"
          }
        }
      ],
      "blockhash": "6dc5ee79d0810e55a22e993b7cef82c69b5608ce678b780cc298ddd66588e43f"
    },
    "index": 0,
    "chainindex": 200,
    "merklebranch": [
      "07712d088e5903a357e1133d20f6896583cbe6b98ba8f1bbe8b176454fbc7d60",
      "354e3f60cf56cc30e048797e3a54a1ee63467fed5ac41fb6ad8e1cc4649fc28b",
      "758a968e81e6b7b874187b2aadd9c92507f63a01f2a53508bc0270494f7ffe45",
      "4da0aab7ac5a57eb2bb3dd1ccc9bcaf9b7314384ee49feff39b2fda930a919b3",
      "2203b6976a287586add7e9fef29d1e8ffa065de7d5793d17d2e0fbc61b665cc8",
      "18bfb8688f6d7e09a96e4b721115fd84bce951047b38256d4f22d070c1581ab0",
      "ba158746c4bf333d5e74a8976da7e7e8c2ecaee5686d750caf09ecedeffe6e17",
      "a738dd18702d88ad67efbac9ef40587143ef2601f5da0d7343bc770399c0cbf6",
      "21a9946bd27fd31f899c49e066fce68505e24b0c070af69ba65334ee57ca8608",
      "32cc99d8e6478967cf458ba81b2ec907d6f0d30da864960942de0c46cdebfb42"
    ],
    "chainmerklebranch": [
      "0000000000000000000000000000000000000000000000000000000000000000",
      "f98c4e9736d8eb8bb46299798906695c755369a3df99a93ffdded1713f1cf6e2",
      "48e55233b9707330def98c80a2105eaa5fac8f687d872ffb4b4741fa2bdb247d",
      "c77d206e2f3cc38e18b85aaa89a8f142a9bf0f4106925d39708f91083e7d8594",
      "6e689e597bf827a6925e702a0a86ea9fdeb2a4167e6172f304c217e3e6e87def",
      "be940f0ee892d50c130caf72dba25372dc6112d97c2ec135753c8b4220ae694a",
      "f3cfed71343ac7d19dba62aed249f748f7fdda194f8d53555dc6a323f6b4d979",
      "55d590281d682e6baf9912a6a9cdf369ce381d4b77266d4c3ede50ec4986b054"
    ],
    "parentblock": "000000203773c13b8ce3cb0043ac33fe15ba85a6bbff5e6a2ac0f6883658b258f1e16665bab4d98f11415131d01319d8d6b1025f0d33a580cd4645c84a9d787e054efc218c132c6670a674190bc98354"
  },
  "previousblockhash": "b9bc4f878376dff79c5a7bae6e66d1a85d9b4828084966c4c6a7882e97a21802",
  "nextblockhash": "3b0a19b917c438b4af0dd0fa5dffb35ec8c38364d4877909ac042a99201c56a7"
}
```

Yeah, there's not much documentation for merge mining unfortunately.

**Comment by pepecoinppc** [comment]
Ok, I've been poking around in the Pepecoin (basically Dogecoin) source code. In `auxpow.cpp` the line `return error("Aux POW missing chain merkle root in parent coinbase");` can be found.

https://github.com/pepecoinppc/pepecoin/blob/master/src/auxpow.cpp#L81-L157

The if statement on line 115 checks to see if something called the `vchRootHash` was contained in parent coinbase (I think that's what vin[0] is?)

This looks like where the magic is happening

![image](https://github.com/dreams-money/merged-mining-pool/assets/143240103/0c2e9a4e-a922-4b3c-918a-aa7256ecc9ca)

**Reply by pepecoinppc** [reply]
If `pc == script.end()` that means that `vchRootHash` is not a substring of `script`. So the solution here seems to be to include the `vchRootHash`.

**Comment by pepecoinppc** [comment]
Ok here's an example of it running the "check" function once. At line 114, I added print statements which print the hex values of vchRootHash and script.

```
vchRootHash: 468d88765922879e83442e69f3fbd61408b2fb96715eae5dc35cd6f57d613e1d
script: 030879314105f9ae01000000464d696e656420627920506570654775696c642e636f6dfabe6d6dba8d3f5a22d58fda39e59190679a6a1e522bcd234999bddf5e00bdc143da685a010000000000000000002632
```

You know more about this than me. I believe this is what it's looking for though. `vchRootHash` needs to be a substring of `script` and it isn't.

```
const uint256 nRootHash
      = CheckMerkleBranch(hashAuxBlock, vChainMerkleBranch, nChainIndex);
    std::vector<unsigned char> vchRootHash(nRootHash.begin(), nRootHash.end());
    std::reverse(vchRootHash.begin(), vchRootHash.end()); // correct endian
```

`vchRootHash` looks to be a function of `hashAuxBlock`, `vChainMerkleBranch`, and `nChainIndex` (then reversed) which I believe are available to the pool? This is where I'm reaching the edge of my competency.

It looks like we're on the right track here though.

**Comment by pepecoinppc** [comment]
Sorry, just trying to document my progress for you.

The TLDR; is that the check function in Dogecoin/Pepecoin here https://github.com/pepecoinppc/pepecoin/blob/master/src/auxpow.cpp#L81-L157 seems to be what you're looking for.

**Comment by pepecoinppc** [comment]
Ok, in the coinbase value, the check function is expecting the following string of hex

`mergedMiningHeader` == `fabe6d6d` (4 bytes)
`vchRootHash` (aka your `auxBlockHash`) (32 bytes)
`merkleHeight` (4 bytes)
`nonce` (4 bytes) - Seems to just be `00000000` everytime?

Here's a Dogecoin block with a coinbase that has these values. This is block 5190783. Mined by Mining Dutch.

0455bf2401046f282f6608 **fabe6d6d** **94c46eb28cf007751829674b1a7fc185572f74e04967e8a09d5fe0e039d1e540** **00010000** **00000000** 01c79a23c30400004603112f4d696e696e672d44757463682f2d3333

Update: Alright, I see this documented in the docs. Doing further research.

**Comment by pepecoinppc** [comment]
Right now, we are using the "Hash of the AuxPOW block header" (aka `auxBlockHash`)

I believe we need to be using the "root of the chain merkle branch". 

![image](https://github.com/dreams-money/merged-mining-pool/assets/143240103/67aa88bc-8ba8-49ab-82ff-b6f42d1a15ce)
[source](https://en.bitcoin.it/wiki/Merged_mining_specification)

Here's an illustration of the parent chain's merkle tree [source](https://tlu.tarilabs.com/mining/MergedMiningIntroduction)

![image](https://github.com/dreams-money/merged-mining-pool/assets/143240103/9ab13b6c-af2a-48cf-bdc3-248482596e1b)

`vchRootHash` should be the hash of the merkle tree root. However, there's an issue, we need to add the auxiliary chains (as illustrated) as leaves _before_ we calculate the merkle root hash. The hash serves as a "proof of inclusion" of the auxiliary chains.

It looks like the `merkle_size` is the number of leaves rounded up to the nearest power of 2. This would be different for each chain.

3 => 4
4 => 4
5 => 8
9 => 16
etc.

I think `merkle_nonce` is always zero. The purpose of this value is to avoid leaf collisions when inserting each aux chain in the parent merkle tree illustrated above, but since it doesn't work, it's up to the child blockchain developers to pick aux chain id's that don't cause leaf collisions.

> The idea is that you can increment merkle_nonce until the chains you're mining don't clash for the same slot. The trouble is that this doesn't work; because it just adds a number derived from the merkle_nonce to the chain_id, if two chains clash for one nonce they'll still clash for all possible nonces.[[1]](https://en.bitcoin.it/wiki/Merged_mining_specification#cite_note-1) New implementers: please pick your chain_id so that not clashing with existing chains requires as small a value of merkle_size as possible, or use a better algorithm to calculate the slot id for your chain.

[source](https://en.bitcoin.it/wiki/Merged_mining_specification)

Does this help? Sorry if it's overwhelming.

**Reply by dreams-money** [reply]
Hey.  Haven't forgot about this.  Been in the hospital with a family emergency. I'll digest all this in the near future.

**Reply by dreams-money** [reply]
Okay.  I'm home now.  I'm downloading a blockchain with a full transaction index.  We'll figure this out.  Are you on Discord?

**Reply by BenzMo** [reply]
Did you manage to figure this out? @pepecoinppc

**Comment by dreams-money** [comment]
The code would need to be updated in a few more places, but the key functions to look at for multi-coin mining are AuxBlock.GetWork(), AuxBlock.MakeAuxPow(), and AuxBlock.makeAuxChainMerkleBranch()

**Comment by mamafunny** [comment]
good evening, how is the 2nd auxpow implement is going on?

**Comment by wombatlabs** [comment]
This pool software supports multiple merged coins if it's of any help:
https://github.com/eXtremal-ik7/poolcore/tree/version/0.4

**Reply by mamafunny** [reply]
are u able to add it into mc?

**Reply by wombatlabs** [reply]
@mamafunny Yes I can

**Reply by headsortales** [reply]
Hi there - I'm also looking for some help on multiple merged coins? Anyone avail to consult? Thanks in advance!

**Reply by wombatlabs** [reply]
@headsortales Yes, shoot me a message. My contact info is on my profile.

**Reply by headsortales** [reply]
> @headsortales Yes, shoot me a message. My contact info is on my profile.

Emailed and messaged on telegram. Looking forward to your reply!

**Comment by Mecanik** [comment]
@pepecoinppc did you manage to make this work? Thanks!

**Comment by gregor-coder** [comment]
@dreams-money I've submitted a PR #22 . Could you please review it when you have a chance?
